include/boost/gil/image_processing/morphology.hpp

```diff
@@ -0,0 +1,193 @@
+//
+// Copyright 2021 Prathamesh Tagore <prathameshtagore@gmail.com>
+//
+// Use, modification and distribution are subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BOOST_GIL_IMAGE_PROCESSING_MORPHOLOGY_HPP
+#define BOOST_GIL_IMAGE_PROCESSING_MORPHOLOGY_HPP
+
+#include <boost/gil/extension/numeric/kernel.hpp>
+
+#include <boost/gil/gray.hpp>
+#include <boost/gil/algorithm.hpp>
+#include <boost/gil/image_view.hpp>
+#include <boost/gil/image_processing/threshold.hpp>
+
+namespace gil = boost::gil;
+namespace boost{
+    namespace gil{
+        enum class morphological_operations
+            {
+                dilate,
+                erode,
+                opening,
+                closing,
+                morphological_gradient,
+                top_hat,
+                black_hat
+            };
+namespace detail{
+template <typename SrcView, typename DstView, typename Kernel>
+void morph_impl(SrcView const& src_view, DstView const& dst_view, Kernel const& kernel,const int identifier)
+{
+    int flip_ker_row, flip_ker_col, row_boundary, col_boundary;
+    typename channel_type<typename SrcView::value_type>::type max_overlapped_element,min_overlapped_element;
+    for (std::ptrdiff_t view_row = 0; view_row < src_view.height(); ++view_row)
+    {
+        for (std::ptrdiff_t view_col = 0; view_col < src_view.width(); ++view_col)
+        {
+            max_overlapped_element = (std::numeric_limits<typename channel_type<SrcView>::type>::min)();
+            min_overlapped_element = (std::numeric_limits<typename channel_type<SrcView>::type>::max)();
+            for (std::size_t kernel_row = 0; kernel_row < kernel.size(); ++kernel_row)
+            {
+                flip_ker_row = kernel.size() - 1 - kernel_row;      // row index of flipped kernel
+
+                for (std::size_t kernel_col = 0; kernel_col < kernel.size(); ++kernel_col)
+                {
+                    flip_ker_col = kernel.size() - 1 - kernel_col; // column index of flipped kernel
+
+                    // index of input signal, used for checking boundary
+                    row_boundary = view_row + (kernel.center_y() - flip_ker_row);
+                    col_boundary = view_col + (kernel.center_x() - flip_ker_col);
+
+                    // ignore input samples which are out of bound
+                    if (row_boundary >= 0 && row_boundary < src_view.height() &&
+                        col_boundary >= 0 && col_boundary < src_view.width())
+                    {
+                        if(src_view(col_boundary, row_boundary) > max_overlapped_element)
+                            max_overlapped_element = src_view(col_boundary, row_boundary);
+                        if(src_view(col_boundary, row_boundary) < min_overlapped_element)
+                            min_overlapped_element = src_view(col_boundary, row_boundary);
```

Comment:
I believe you forgot to multiply by kernel value at that point.

I believe you should create a function pointer before entering those loops, so that only one element is searched for (e.g. either min or max, but not both). The easiest option I see is to accept a functor that choose the right element by changing the declaration to this:

```
template <typename SrcView, typename DstView, typename Kernel, typename SelectionFunctor>
void morph_impl(SrcView const& src_view, DstView const& dst_view, Kernel const& kernel, SelectionFunctor selector)
```

Then the four lines will be reduced to this:

```
    target_element = selector(target_element, src_view(col_boundary, row_boundary));
```

Then just `switch` inside `morph` to pass correct functor. Remember, you have to pass `&std::max<channel_type>` or `&std::min<channel_type>`, otherwise you cannot take a pointer to template function, you have to explicitly tell which version you need.

Reply:
Yeah my bad, I forgot to multiply.
Regarding the second point,can we solve it with simple if else conditions like
```
 product = src_view(col_boundary, row_boundary) * kernel.at(flip_ker_row, flip_ker_col);
                        if(identifier == 1)
                        {
                            if(product > max_overlapped_element)
                                max_overlapped_element = src_view(col_boundary, row_boundary);
                        }
                        else if(identifier == 0)
                        {
                            if(product < min_overlapped_element)
                                min_overlapped_element = src_view(col_boundary, row_boundary);
                        }
```
or is there any other specific reason for using functors, I am fine with either approaches

Reply:
To me, one of reasons to prefer functor or named lambda is readability:
- here we search max, there we search min, and I don't have to look further to find the branching, and decoded it in my brain when happens what
- `if(identifier == 1)` is just a terribly obscure: wtf is the `identifier`, wtf is the magic `1` - how on earth the magic `identifier == 1` corresponds to searching for min/max?!

By he way, you create an enum earlier, why you are refusing to use it and preferring the magic numbers?
Please, write code for humans first, then for machines!

Reply:
@mloskot ,got the point

Reply:
@meshtag The message I was trying to carry through in the chatty way was that if we write code [Sean Parent's 1st goal](https://www.youtube.com/watch?v=W2tWOdzgXHA) in mind for [better code](https://www.youtube.com/results?search_query=sean+parent+better+code), making intentions more explicit and well named, the life would be easier :-)

Reply:
Okay, thanks @mloskot ,I will try to name my variables and functions in a more expressive sense from now on.

Reply:
@simmplecoder ,
I am having trouble figuring out what exactly you wanted me to do here 
``` Remember, you have to pass &std::max<channel_type> or &std::min<channel_type>```
can u please elaborate ,I understand we cannot have a pointer to a template and template is the most suitable way for our function to handle different types of gil views

Reply:
@simmplecoder ,
In my recent commit , I have handled this part as 
```
if(identifier == boost::gil::morphological_operations::dilation)
                            {
                                if(src_view(col_boundary, row_boundary) > target_element)
                                    target_element = src_view(col_boundary, row_boundary);
                            }
                            else if(identifier == boost::gil::morphological_operations::erosion)
                            {
                                if(src_view(col_boundary, row_boundary) < target_element)
                                    target_element = src_view(col_boundary, row_boundary);
                            }
```
since I was not sure regarding the implementation of functor you suggested.Please check this part and tell me if I need to improve it further.

Reply:
@meshtag, I will send you a patch regarding the function switch depending on the erosion/dilation mode. I will resolve this thread for now, good job.